pstoreds/addr_book.go

```diff
@@ -2,398 +2,338 @@ package pstoreds
+//    permanent, popular values used in other libp2p modules. In this cited case, optimizing with lookahead windows
+//    makes little sense.
+func NewAddrBook(ctx context.Context, store ds.Batching, opts Options) (ab *dsAddrBook, err error) {
+	var cache cache = new(noopCache)
+	if opts.CacheSize > 0 {
+		if cache, err = lru.NewARC(int(opts.CacheSize)); err != nil {
+			return nil, err
+		}
 	}
 
-	mgr.cache.Remove(p)
-	// Attempt transactional KV deletion.
-	for i := 0; i < mgr.writeRetries; i++ {
-		if err = mgr.dbDelete(keys); err == nil {
-			break
-		}
-		log.Errorf("failed to delete addresses for peer %s: %s\n", p.Pretty(), err)
+	ctx, cancelFn := context.WithCancel(ctx)
+	ab = &dsAddrBook{
+		ctx:         ctx,
+		cancelFn:    cancelFn,
+		opts:        opts,
+		cache:       cache,
+		ds:          store,
+		subsManager: pstoremem.NewAddrSubManager(),
+		flushJobCh:  make(chan *addrsRecord, 32),
 	}
 
-	if err != nil {
-		log.Errorf("failed to avoid write conflict for peer %s after %d retries: %v\n", p.Pretty(), mgr.writeRetries, err)
-		return err
+	if ab.gc, err = newAddressBookGc(ctx, ab); err != nil {
+		return nil, err
 	}
 
-	return nil
-}
+	go ab.flusher()
 
-func (mgr *dsAddrBook) setAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration, mode ttlWriteMode) error {
-	// Keys and cleaned up addresses.
-	keys, addrs, err := keysAndAddrs(p, addrs)
-	if err != nil {
-		return err
-	}
+	return ab, nil
+}
 
-	mgr.cache.Remove(p)
-	// Attempt transactional KV insertion.
-	var existed []bool
-	for i := 0; i < mgr.writeRetries; i++ {
-		if existed, err = mgr.dbInsert(keys, addrs, ttl, mode); err == nil {
-			break
-		}
-		log.Errorf("failed to write addresses for peer %s: %s\n", p.Pretty(), err)
-	}
+func (ab *dsAddrBook) Close() {
+	ab.cancelFn()
+	ab.done.Wait()
+}
 
-	if err != nil {
-		log.Errorf("failed to avoid write conflict for peer %s after %d retries: %v\n", p.Pretty(), mgr.writeRetries, err)
-		return err
+func (ab *dsAddrBook) asyncFlush(pr *addrsRecord) {
+	select {
+	case ab.flushJobCh <- pr:
+	default:
+		log.Warningf("flush queue is full; could not flush record for peer %s", pr.Id.ID)
 	}
+}
 
-	// Update was successful, so broadcast event only for new addresses.
-	for i, _ := range keys {
-		if !existed[i] {
-			mgr.subsManager.BroadcastAddr(p, addrs[i])
+// loadRecord is a read-through fetch. It fetches a record from cache, falling back to the
+// datastore upon a miss, and returning a newly initialized record if the peer doesn't exist.
+//
+// loadRecord calls clean() on existing recordsrecord before returning it. If the record changes
+// as a result and the update argument is true, an async flush is queued.
+//
+// If the cache argument is true, the record is inserted in the cache when loaded from the datastore.
+func (ab *dsAddrBook) loadRecord(id peer.ID, cache bool, update bool) (pr *addrsRecord, err error) {
+	if e, ok := ab.cache.Get(id); ok {
+		pr = e.(*addrsRecord)
```

Comment:
I know it should not fail the assertion, but I would prefer if we checked it here. It is small change but much better than a panic.

Reply:
Really? I'd rather panic if we have a bug.

Reply:
I would rather not crash the user's instance and log it as an error. It shouldn't happen either way but is possible to recover completely from an issue like that, then why should we panic?

IMO in a grade of software, we are creating panics have to reasons to exist: not possible to return an error (Unsafe creators and so on) and errors that we cannot recover from.

Reply:
This cannot happen unless memory is corrupted (runtime) or a change in code ends up storing a wrong type (development time). In my view, both cases warrant panicking right here right now, rather than swallowing the error or removing the record with a warning that might never get noticed.

Reply:
If this was being loaded from a disk store, I’d definitely catch it, as corruption could have many causes. But if memory is gets corrupted, that’s an indication of a deeper problem and I’d rather abort entirely.